[BUG] [Segment Replication] Shard failures on node stop/restart

When a node (say Node A) containing primary with on going replication leave the cluster; it results in replica shard failure on target (Node B) due to `NodeClosedException`. For one replica count, this also leads to a red cluster because both primary (on node A) and replica (on node B) are unassigned. This can be resolved by handling the exceptions gracefully on target when node leaves the cluster. 

```
[2023-03-07T22:11:37,498][ERROR][o.o.i.r.SegmentReplicationTargetService] [ip-10-0-4-54.us-west-2.compute.internal] replication failure
org.opensearch.indices.replication.common.ReplicationFailedException: Segment Replication failed
        at org.opensearch.indices.replication.SegmentReplicationTargetService$3.onFailure(SegmentReplicationTargetService.java:365) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListener$1.onFailure(ActionListener.java:88) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionRunnable.onFailure(ActionRunnable.java:103) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:54) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.OpenSearchExecutors$DirectExecutorService.execute(OpenSearchExecutors.java:343) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.ListenableFuture.notifyListener(ListenableFuture.java:120) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.ListenableFuture.lambda$done$0(ListenableFuture.java:112) [opensearch-2.7.0.jar:2.7.0]
        at java.util.ArrayList.forEach(ArrayList.java:1511) [?:?]
        at org.opensearch.common.util.concurrent.ListenableFuture.done(ListenableFuture.java:112) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.BaseFuture.setException(BaseFuture.java:178) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.ListenableFuture.onFailure(ListenableFuture.java:149) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.StepListener.innerOnFailure(StepListener.java:82) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.NotifyOnceListener.onFailure(NotifyOnceListener.java:62) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListener$4.onFailure(ActionListener.java:190) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListener$6.onFailure(ActionListener.java:309) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.support.RetryableAction$RetryingListener.onFinalFailure(RetryableAction.java:218) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.support.RetryableAction$RetryingListener.onFailure(RetryableAction.java:210) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:74) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1414) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.transport.InboundHandler.lambda$handleException$3(InboundHandler.java:420) [opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:747) [opensearch-2.7.0.jar:2.7.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
        at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: org.opensearch.transport.RemoteTransportException: [ip-10-0-5-14.us-west-2.compute.internal][10.0.5.14:9300][internal:index/shard/replication/get_segment_files]
Caused by: org.opensearch.transport.SendRequestTransportException: [ip-10-0-4-54.us-west-2.compute.internal][10.0.4.54:9300][internal:index/shard/replication/file_chunk]
        at org.opensearch.transport.TransportService.sendRequestInternal(TransportService.java:941) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.transport.TransportService.sendRequest(TransportService.java:815) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.transport.TransportService.sendRequest(TransportService.java:758) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.recovery.RetryableTransportClient$1.tryAction(RetryableTransportClient.java:91) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.support.RetryableAction$1.doRun(RetryableAction.java:137) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.OpenSearchExecutors$DirectExecutorService.execute(OpenSearchExecutors.java:343) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.support.RetryableAction.run(RetryableAction.java:115) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.recovery.RetryableTransportClient.executeRetryableAction(RetryableTransportClient.java:106) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.replication.RemoteSegmentFileChunkWriter.writeFileChunk(RemoteSegmentFileChunkWriter.java:117) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.replication.SegmentFileTransferHandler$1.executeChunkRequest(SegmentFileTransferHandler.java:148) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.replication.SegmentFileTransferHandler$1.executeChunkRequest(SegmentFileTransferHandler.java:97) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.recovery.MultiChunkTransfer.handleItems(MultiChunkTransfer.java:149) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.recovery.MultiChunkTransfer$1.write(MultiChunkTransfer.java:98) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.AsyncIOProcessor.processList(AsyncIOProcessor.java:129) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.AsyncIOProcessor.drainAndProcessAndRelease(AsyncIOProcessor.java:117) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.AsyncIOProcessor.put(AsyncIOProcessor.java:98) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.recovery.MultiChunkTransfer.addItem(MultiChunkTransfer.java:109) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.recovery.MultiChunkTransfer.lambda$handleItems$3(MultiChunkTransfer.java:151) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListener$1.onResponse(ActionListener.java:80) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListener$6.onResponse(ActionListener.java:299) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListener$4.onResponse(ActionListener.java:180) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListener$6.onResponse(ActionListener.java:299) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.support.RetryableAction$RetryingListener.onResponse(RetryableAction.java:181) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListenerResponseHandler.handleResponse(ActionListenerResponseHandler.java:69) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1404) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.transport.InboundHandler.doHandleResponse(InboundHandler.java:393) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.transport.InboundHandler.lambda$handleResponse$1(InboundHandler.java:387) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:747) ~[opensearch-2.7.0.jar:2.7.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: org.opensearch.node.NodeClosedException: node closed {ip-10-0-5-14.us-west-2.compute.internal}{TfIe0XASSY-qM9pY1gCmow}{mdJAYWGfSJmU13TV20PaMA}{10.0.5.14}{10.0.5.14:9300}{di}{shard_indexing_pressure_enabled=true}
        at org.opensearch.transport.TransportService.sendRequestInternal(TransportService.java:922) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.transport.TransportService.sendRequest(TransportService.java:815) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.transport.TransportService.sendRequest(TransportService.java:758) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.recovery.RetryableTransportClient$1.tryAction(RetryableTransportClient.java:91) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.support.RetryableAction$1.doRun(RetryableAction.java:137) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.OpenSearchExecutors$DirectExecutorService.execute(OpenSearchExecutors.java:343) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.support.RetryableAction.run(RetryableAction.java:115) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.recovery.RetryableTransportClient.executeRetryableAction(RetryableTransportClient.java:106) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.replication.RemoteSegmentFileChunkWriter.writeFileChunk(RemoteSegmentFileChunkWriter.java:117) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.replication.SegmentFileTransferHandler$1.executeChunkRequest(SegmentFileTransferHandler.java:148) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.replication.SegmentFileTransferHandler$1.executeChunkRequest(SegmentFileTransferHandler.java:97) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.recovery.MultiChunkTransfer.handleItems(MultiChunkTransfer.java:149) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.recovery.MultiChunkTransfer$1.write(MultiChunkTransfer.java:98) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.AsyncIOProcessor.processList(AsyncIOProcessor.java:129) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.AsyncIOProcessor.drainAndProcessAndRelease(AsyncIOProcessor.java:117) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.AsyncIOProcessor.put(AsyncIOProcessor.java:98) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.recovery.MultiChunkTransfer.addItem(MultiChunkTransfer.java:109) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.indices.recovery.MultiChunkTransfer.lambda$handleItems$3(MultiChunkTransfer.java:151) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListener$1.onResponse(ActionListener.java:80) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListener$6.onResponse(ActionListener.java:299) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListener$4.onResponse(ActionListener.java:180) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListener$6.onResponse(ActionListener.java:299) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.support.RetryableAction$RetryingListener.onResponse(RetryableAction.java:181) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.action.ActionListenerResponseHandler.handleResponse(ActionListenerResponseHandler.java:69) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1404) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.transport.InboundHandler.doHandleResponse(InboundHandler.java:393) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.transport.InboundHandler.lambda$handleResponse$1(InboundHandler.java:387) ~[opensearch-2.7.0.jar:2.7.0]
        at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:747) ~[opensearch-2.7.0.jar:2.7.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]

[2023-03-07T22:11:37,878][INFO ][o.o.c.r.a.AllocationService] [seed] Cluster health status changed from [YELLOW] to [RED] (reason: [shards failed [[nyc_taxis][20], [nyc_taxis][20]]]).
```

### Repro steps
1. Create a multi node cluster with large shard count
2. Stop opensearch process on one node while there is heavy indexing on-going (works well with `nyc_taxis` OpenSearch-Benchmark). The more number of shards on stopped node, the more chances of an on-going replication event. This results in cluster going red for one replica setup.
3. Cluster manager brings the primary up on Node B which contains previously copied files

### Expected
Shard should not be marked failed on target when node containing primary goes down.

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationBaseIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationBaseIT.java
@@ -164,43 +164,6 @@ public class SegmentReplicationBaseIT extends OpenSearchIntegTestCase {
         }, 1, TimeUnit.MINUTES);
     }
 
-    protected void corruptRandomFile(String indexName, String nodeId) throws IOException {
-        ClusterState state = client().admin().cluster().prepareState().get().getState();
-        Index test = state.metadata().index(indexName).getIndex();
-
-        final IndexShard indexShard = getIndexShard(nodeId, INDEX_NAME);
-        final ShardRouting shardRouting = indexShard.routingEntry();
-
-        NodesStatsResponse nodeStats = client().admin().cluster().prepareNodesStats(nodeId).addMetric(FS.metricName()).get();
-        Set<Path> files = new TreeSet<>();
-        for (FsInfo.Path info : nodeStats.getNodes().get(0).getFs()) {
-            String path = info.getPath();
-            Path file = PathUtils.get(path)
-                .resolve("indices")
-                .resolve(test.getUUID())
-                .resolve(Integer.toString(shardRouting.getId()))
-                .resolve("index");
-            if (Files.exists(file)) { // multi data path might only have one path in use
-                try (Directory dir = FSDirectory.open(file)) {
-                    SegmentInfos segmentCommitInfos = Lucene.readSegmentInfos(dir);
-                    for (SegmentCommitInfo commitInfo : segmentCommitInfos) {
-                        if (commitInfo.getDelCount() + commitInfo.getSoftDelCount() == commitInfo.info.maxDoc()) {
-                            // don't corrupt fully deleted segments - they might be removed on snapshot
-                            continue;
-                        }
-                        for (String commitFile : commitInfo.files()) {
-                            if (commitFile.contains("segments") == false) {
-                                files.add(file.resolve(commitFile));
-                            }
-                            logger.info("--> File name {}", commitFile);
-                        }
-                    }
-                }
-            }
-        }
-        CorruptionUtils.corruptFile(random(), files.toArray(new Path[0]));
-    }
-
     protected void verifyStoreContent() throws Exception {
         assertBusy(() -> {
             final ClusterState clusterState = getClusterState();

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationBaseIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationBaseIT.java
@@ -8,11 +8,6 @@
 
 package org.opensearch.indices.replication;
 
-import org.apache.lucene.index.SegmentCommitInfo;
-import org.apache.lucene.index.SegmentInfos;
-import org.apache.lucene.store.Directory;
-import org.apache.lucene.store.FSDirectory;
-import org.opensearch.action.admin.cluster.node.stats.NodesStatsResponse;
 import org.opensearch.action.admin.indices.replication.SegmentReplicationStatsResponse;
 import org.opensearch.action.search.SearchResponse;
 import org.opensearch.cluster.ClusterState;
@@ -22,8 +17,6 @@ import org.opensearch.cluster.routing.IndexRoutingTable;
 import org.opensearch.cluster.routing.IndexShardRoutingTable;
 import org.opensearch.cluster.routing.ShardRouting;
 import org.opensearch.common.Nullable;
-import org.opensearch.common.io.PathUtils;
-import org.opensearch.common.lucene.Lucene;
 import org.opensearch.common.settings.Settings;
 import org.opensearch.common.util.FeatureFlags;
 import org.opensearch.index.Index;
@@ -34,27 +27,19 @@ import org.opensearch.index.store.Store;
 import org.opensearch.index.store.StoreFileMetadata;
 import org.opensearch.indices.IndicesService;
 import org.opensearch.indices.replication.common.ReplicationType;
-import org.opensearch.monitor.fs.FsInfo;
 import org.opensearch.plugins.Plugin;
-import org.opensearch.test.CorruptionUtils;
 import org.opensearch.test.OpenSearchIntegTestCase;
 import org.opensearch.test.transport.MockTransportService;
 
-import java.io.IOException;
-import java.nio.file.Files;
-import java.nio.file.Path;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
-import java.util.Set;
-import java.util.TreeSet;
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 
 import static java.util.Arrays.asList;
-import static org.opensearch.action.admin.cluster.node.stats.NodesStatsRequest.Metric.FS;
 import static org.opensearch.test.OpenSearchIntegTestCase.client;
 import static org.opensearch.test.OpenSearchTestCase.assertBusy;
 import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount;

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationBaseIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationBaseIT.java
@@ -8,6 +8,12 @@
 
 package org.opensearch.indices.replication;
 
+import org.apache.lucene.index.SegmentCommitInfo;
+import org.apache.lucene.index.SegmentInfos;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FSDirectory;
+import org.opensearch.action.admin.cluster.node.stats.NodesStatsResponse;
+import org.opensearch.action.admin.indices.replication.SegmentReplicationStatsResponse;
 import org.opensearch.action.search.SearchResponse;
 import org.opensearch.cluster.ClusterState;
 import org.opensearch.cluster.metadata.IndexMetadata;
@@ -16,6 +22,8 @@ import org.opensearch.cluster.routing.IndexRoutingTable;
 import org.opensearch.cluster.routing.IndexShardRoutingTable;
 import org.opensearch.cluster.routing.ShardRouting;
 import org.opensearch.common.Nullable;
+import org.opensearch.common.io.PathUtils;
+import org.opensearch.common.lucene.Lucene;
 import org.opensearch.common.settings.Settings;
 import org.opensearch.common.util.FeatureFlags;
 import org.opensearch.index.Index;
@@ -26,19 +34,27 @@ import org.opensearch.index.store.Store;
 import org.opensearch.index.store.StoreFileMetadata;
 import org.opensearch.indices.IndicesService;
 import org.opensearch.indices.replication.common.ReplicationType;
+import org.opensearch.monitor.fs.FsInfo;
 import org.opensearch.plugins.Plugin;
+import org.opensearch.test.CorruptionUtils;
 import org.opensearch.test.OpenSearchIntegTestCase;
 import org.opensearch.test.transport.MockTransportService;
 
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
+import java.util.TreeSet;
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 
 import static java.util.Arrays.asList;
+import static org.opensearch.action.admin.cluster.node.stats.NodesStatsRequest.Metric.FS;
 import static org.opensearch.test.OpenSearchIntegTestCase.client;
 import static org.opensearch.test.OpenSearchTestCase.assertBusy;
 import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount;
@@ -132,6 +148,57 @@ public class SegmentReplicationBaseIT extends OpenSearchIntegTestCase {
 
     protected void waitForSearchableDocs(long docCount, String... nodes) throws Exception {
         waitForSearchableDocs(docCount, Arrays.stream(nodes).collect(Collectors.toList()));
+    }
+
+    protected void waitForSegmentReplication(String node) throws Exception {
+        assertBusy(() -> {
+            SegmentReplicationStatsResponse segmentReplicationStatsResponse = client(node).admin()
+                .indices()
+                .prepareSegmentReplicationStats(INDEX_NAME)
+                .execute()
+                .actionGet();
+            assertEquals(
+                segmentReplicationStatsResponse.shardSegmentReplicationStates().get(INDEX_NAME).get(0).getStage(),
+                SegmentReplicationState.Stage.DONE
+            );
+        }, 1, TimeUnit.MINUTES);
+    }
+
+    protected void corruptRandomFile(String indexName, String nodeId) throws IOException {
+        ClusterState state = client().admin().cluster().prepareState().get().getState();
+        Index test = state.metadata().index(indexName).getIndex();
+
+        final IndexShard indexShard = getIndexShard(nodeId, INDEX_NAME);
+        final ShardRouting shardRouting = indexShard.routingEntry();
+
+        NodesStatsResponse nodeStats = client().admin().cluster().prepareNodesStats(nodeId).addMetric(FS.metricName()).get();
+        Set<Path> files = new TreeSet<>();
+        for (FsInfo.Path info : nodeStats.getNodes().get(0).getFs()) {
+            String path = info.getPath();
+            Path file = PathUtils.get(path)
+                .resolve("indices")
+                .resolve(test.getUUID())
+                .resolve(Integer.toString(shardRouting.getId()))
+                .resolve("index");
+            if (Files.exists(file)) { // multi data path might only have one path in use
+                try (Directory dir = FSDirectory.open(file)) {
+                    SegmentInfos segmentCommitInfos = Lucene.readSegmentInfos(dir);
+                    for (SegmentCommitInfo commitInfo : segmentCommitInfos) {
+                        if (commitInfo.getDelCount() + commitInfo.getSoftDelCount() == commitInfo.info.maxDoc()) {
+                            // don't corrupt fully deleted segments - they might be removed on snapshot
+                            continue;
+                        }
+                        for (String commitFile : commitInfo.files()) {
+                            if (commitFile.contains("segments") == false) {
+                                files.add(file.resolve(commitFile));
+                            }
+                            logger.info("--> File name {}", commitFile);
+                        }
+                    }
+                }
+            }
+        }
+        CorruptionUtils.corruptFile(random(), files.toArray(new Path[0]));
     }
 
     protected void verifyStoreContent() throws Exception {

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
@@ -9,11 +9,6 @@
 package org.opensearch.indices.replication;
 
 import com.carrotsearch.randomizedtesting.RandomizedTest;
-import org.apache.lucene.index.SegmentCommitInfo;
-import org.apache.lucene.index.SegmentInfos;
-import org.apache.lucene.store.Directory;
-import org.apache.lucene.store.FSDirectory;
-import org.opensearch.action.admin.cluster.node.stats.NodesStatsResponse;
 import org.opensearch.action.support.WriteRequest;
 import org.opensearch.action.update.UpdateResponse;
 import org.opensearch.client.Requests;
@@ -23,36 +18,25 @@ import org.opensearch.cluster.metadata.IndexMetadata;
 import org.opensearch.cluster.node.DiscoveryNode;
 import org.opensearch.cluster.routing.ShardRouting;
 import org.opensearch.cluster.routing.allocation.command.CancelAllocationCommand;
-import org.opensearch.common.io.PathUtils;
-import org.opensearch.common.lucene.Lucene;
 import org.opensearch.common.settings.Settings;
-import org.opensearch.index.Index;
 import org.opensearch.index.IndexModule;
 import org.opensearch.index.shard.IndexShard;
 import org.opensearch.indices.recovery.FileChunkRequest;
 import org.opensearch.indices.replication.common.ReplicationType;
-import org.opensearch.monitor.fs.FsInfo;
 import org.opensearch.node.NodeClosedException;
 import org.opensearch.test.BackgroundIndexer;
-import org.opensearch.test.CorruptionUtils;
 import org.opensearch.test.InternalTestCluster;
 import org.opensearch.test.OpenSearchIntegTestCase;
 import org.opensearch.test.transport.MockTransportService;
 import org.opensearch.transport.TransportService;
 
-import java.io.IOException;
-import java.nio.file.Files;
-import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Set;
-import java.util.TreeSet;
 import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.TimeUnit;
 
 import static java.util.Arrays.asList;
-import static org.opensearch.action.admin.cluster.node.stats.NodesStatsRequest.Metric.FS;
 import static org.opensearch.index.query.QueryBuilders.matchQuery;
 import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
 import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount;
@@ -383,7 +367,14 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
         internalCluster().startClusterManagerOnlyNode();
         final String primaryNode = internalCluster().startNode();
         // create index without auto refreshes
-        createIndex(INDEX_NAME, Settings.builder().put(indexSettings()).put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 1).put("index.refresh_interval", -1).build());
+        createIndex(
+            INDEX_NAME,
+            Settings.builder()
+                .put(indexSettings())
+                .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 1)
+                .put("index.refresh_interval", -1)
+                .build()
+        );
         ensureYellow(INDEX_NAME);
         final String replicaNode = internalCluster().startNode();
         ensureGreen(INDEX_NAME);

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
@@ -9,16 +9,6 @@
 package org.opensearch.indices.replication;
 
 import com.carrotsearch.randomizedtesting.RandomizedTest;
-import org.apache.lucene.document.Document;
-import org.apache.lucene.document.Field;
-import org.apache.lucene.document.SortedDocValuesField;
-import org.apache.lucene.document.StringField;
-import org.apache.lucene.document.TextField;
-import org.apache.lucene.index.IndexWriter;
-import org.apache.lucene.index.IndexWriterConfig;
-import org.apache.lucene.index.SegmentInfos;
-import org.apache.lucene.tests.util.TestUtil;
-import org.apache.lucene.util.BytesRef;
 import org.opensearch.action.support.WriteRequest;
 import org.opensearch.action.update.UpdateResponse;
 import org.opensearch.client.Requests;
@@ -30,7 +20,6 @@ import org.opensearch.cluster.routing.ShardRouting;
 import org.opensearch.cluster.routing.allocation.command.CancelAllocationCommand;
 import org.opensearch.common.settings.Settings;
 import org.opensearch.index.IndexModule;
-import org.opensearch.index.mapper.ParseContext;
 import org.opensearch.index.shard.IndexShard;
 import org.opensearch.indices.recovery.FileChunkRequest;
 import org.opensearch.indices.replication.common.ReplicationType;
@@ -42,7 +31,6 @@ import org.opensearch.test.transport.MockTransportService;
 import org.opensearch.transport.TransportService;
 
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.List;
 import java.util.Set;
 import java.util.concurrent.CountDownLatch;

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
@@ -9,6 +9,11 @@
 package org.opensearch.indices.replication;
 
 import com.carrotsearch.randomizedtesting.RandomizedTest;
+import org.apache.lucene.index.SegmentCommitInfo;
+import org.apache.lucene.index.SegmentInfos;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FSDirectory;
+import org.opensearch.action.admin.cluster.node.stats.NodesStatsResponse;
 import org.opensearch.action.support.WriteRequest;
 import org.opensearch.action.update.UpdateResponse;
 import org.opensearch.client.Requests;
@@ -18,24 +23,35 @@ import org.opensearch.cluster.metadata.IndexMetadata;
 import org.opensearch.cluster.node.DiscoveryNode;
 import org.opensearch.cluster.routing.ShardRouting;
 import org.opensearch.cluster.routing.allocation.command.CancelAllocationCommand;
+import org.opensearch.common.io.PathUtils;
+import org.opensearch.common.lucene.Lucene;
 import org.opensearch.common.settings.Settings;
+import org.opensearch.index.Index;
 import org.opensearch.index.IndexModule;
 import org.opensearch.index.shard.IndexShard;
 import org.opensearch.indices.recovery.FileChunkRequest;
 import org.opensearch.indices.replication.common.ReplicationType;
+import org.opensearch.monitor.fs.FsInfo;
 import org.opensearch.node.NodeClosedException;
 import org.opensearch.test.BackgroundIndexer;
+import org.opensearch.test.CorruptionUtils;
 import org.opensearch.test.InternalTestCluster;
 import org.opensearch.test.OpenSearchIntegTestCase;
 import org.opensearch.test.transport.MockTransportService;
 import org.opensearch.transport.TransportService;
 
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import java.util.Set;
+import java.util.TreeSet;
 import java.util.concurrent.CountDownLatch;
 
 import static java.util.Arrays.asList;
+import static org.opensearch.action.admin.cluster.node.stats.NodesStatsRequest.Metric.FS;
 import static org.opensearch.index.query.QueryBuilders.matchQuery;
 import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
 import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount;
@@ -359,6 +375,91 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
         blockFileCopy.countDown();
         internalCluster().stopRandomNode(InternalTestCluster.nameFilter(primaryNode));
         assertBusy(() -> { assertDocCounts(docCount, replicaNode); });
+    }
+
+    protected ShardRouting corruptRandomFile(String indexName, String nodeId) throws IOException {
+//        ClusterState state = client().admin().cluster().prepareState().get().getState();
+//        Index test = state.metadata().index("test").getIndex();
+//        GroupShardsIterator shardIterators = state.getRoutingTable().activePrimaryShardsGrouped(new String[] { "test" }, false);
+//        List<ShardIterator> iterators = iterableAsArrayList(shardIterators);
+//        ShardIterator shardIterator = RandomPicks.randomFrom(random(), iterators);
+//        ShardRouting shardRouting = shardIterator.nextOrNull();
+//        assertNotNull(shardRouting);
+//        assertTrue(shardRouting.primary());
+//        assertTrue(shardRouting.assignedToNode());
+//        String nodeId = shardRouting.currentNodeId();
+        ClusterState state = client().admin().cluster().prepareState().get().getState();
+        Index test = state.metadata().index(indexName).getIndex();
+
+        final IndexShard indexShard = getIndexShard(nodeId, INDEX_NAME);
+        final ShardRouting shardRouting = indexShard.routingEntry();
+
+        NodesStatsResponse nodeStatses = client().admin().cluster().prepareNodesStats(nodeId).addMetric(FS.metricName()).get();
+        Set<Path> files = new TreeSet<>(); // treeset makes sure iteration order is deterministic
+        for (FsInfo.Path info : nodeStatses.getNodes().get(0).getFs()) {
+            String path = info.getPath();
+            Path file = PathUtils.get(path)
+                .resolve("indices")
+                .resolve(test.getUUID())
+                .resolve(Integer.toString(shardRouting.getId()))
+                .resolve("index");
+            if (Files.exists(file)) { // multi data path might only have one path in use
+                try (Directory dir = FSDirectory.open(file)) {
+                    SegmentInfos segmentCommitInfos = Lucene.readSegmentInfos(dir);
+//                    if (includePerCommitFiles) {
+//                        files.add(file.resolve(segmentCommitInfos.getSegmentsFileName()));
+//                    }
+                    for (SegmentCommitInfo commitInfo : segmentCommitInfos) {
+                        if (commitInfo.getDelCount() + commitInfo.getSoftDelCount() == commitInfo.info.maxDoc()) {
+                            // don't corrupt fully deleted segments - they might be removed on snapshot
+                            continue;
+                        }
+                        for (String commitFile : commitInfo.files()) {
+//                            if (includePerCommitFiles || isPerSegmentFile(commitFile)) {
+//                                files.add(file.resolve(commitFile));
+//                            }
+                            logger.info("--> File name {}", commitFile);
+                        }
+                    }
+                }
+            }
+        }
+        CorruptionUtils.corruptFile(random(), files.toArray(new Path[0]));
+        return shardRouting;
+    }
+    public void testFileCorruption() throws Exception {
+        internalCluster().startClusterManagerOnlyNode();
+        final String primaryNode = internalCluster().startNode();
+        // create index without auto refreshes
+        createIndex(INDEX_NAME, Settings.builder().put(indexSettings()).put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 1).put("index.refresh_interval", -1).build());
+        ensureYellow(INDEX_NAME);
+        final String replicaNode = internalCluster().startNode();
+        ensureGreen(INDEX_NAME);
+
+        final int initialDocCount = scaledRandomIntBetween(10, 200);
+        for (int i = 0; i < initialDocCount; i++) {
+            client().prepareIndex(INDEX_NAME).setId(String.valueOf(i)).setSource("foo", "bar").get();
+        }
+        logger.info("--> trigger round of segment replication so that replica have files");
+        flush(INDEX_NAME);
+        waitForSearchableDocs(initialDocCount, primaryNode, replicaNode);
+        verifyStoreContent();
+
+        // Primary & replica should be in sync now. Corrupt one file on replica to ensure there is replica shard failure
+        final IndexShard primaryShard = getIndexShard(primaryNode, INDEX_NAME);
+        logger.info("--> Store files {}", Arrays.toString(primaryShard.store().directory().listAll()));
+//        final ShardRouting replicaShardRouting = replicaShard.routingEntry();
+
+//        corruptRandomFile(INDEX_NAME, replicaNode);
+//
+//        // Insert more docs and trigger another round of segment replication
+//        for (int i = initialDocCount; i < 2 * initialDocCount; i++) {
+//            client().prepareIndex(INDEX_NAME).setId(String.valueOf(i)).setSource("foo", "bar").get();
+//        }
+//        logger.info("--> refresh again and expect shard failure on replica");
+//        refresh(INDEX_NAME);
+
+        // Should see segrep failure
     }
 
     public void testCancellation() throws Exception {

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
@@ -324,14 +324,12 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
                 .put(indexSettings())
                 .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
                 .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 1)
+                .put("index.refresh_interval", -1)
                 .build()
         );
         ensureYellow(INDEX_NAME);
         final String replicaNode = internalCluster().startNode();
         ensureGreen(INDEX_NAME);
-
-        // block auto refreshes
-        assertAcked(prepareCreate(INDEX_NAME).setSettings(Settings.builder().put("index.refresh_interval", -1)));
         ClusterState state = client().admin().cluster().prepareState().execute().actionGet().getState();
         // Get replica allocation id
         final String replicaAllocationId = state.routingTable()

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
@@ -9,6 +9,16 @@
 package org.opensearch.indices.replication;
 
 import com.carrotsearch.randomizedtesting.RandomizedTest;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.Field;
+import org.apache.lucene.document.SortedDocValuesField;
+import org.apache.lucene.document.StringField;
+import org.apache.lucene.document.TextField;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.index.IndexWriterConfig;
+import org.apache.lucene.index.SegmentInfos;
+import org.apache.lucene.tests.util.TestUtil;
+import org.apache.lucene.util.BytesRef;
 import org.opensearch.action.support.WriteRequest;
 import org.opensearch.action.update.UpdateResponse;
 import org.opensearch.client.Requests;
@@ -20,6 +30,7 @@ import org.opensearch.cluster.routing.ShardRouting;
 import org.opensearch.cluster.routing.allocation.command.CancelAllocationCommand;
 import org.opensearch.common.settings.Settings;
 import org.opensearch.index.IndexModule;
+import org.opensearch.index.mapper.ParseContext;
 import org.opensearch.index.shard.IndexShard;
 import org.opensearch.indices.recovery.FileChunkRequest;
 import org.opensearch.indices.replication.common.ReplicationType;
@@ -360,50 +371,6 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
         blockFileCopy.countDown();
         internalCluster().stopRandomNode(InternalTestCluster.nameFilter(primaryNode));
         assertBusy(() -> { assertDocCounts(docCount, replicaNode); });
-    }
-
-    @AwaitsFix(bugUrl = "https://github.com/opensearch-project/OpenSearch/issues/6578")
-    public void testFileCorruption() throws Exception {
-        internalCluster().startClusterManagerOnlyNode();
-        final String primaryNode = internalCluster().startNode();
-        // create index without auto refreshes
-        createIndex(
-            INDEX_NAME,
-            Settings.builder()
-                .put(indexSettings())
-                .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 1)
-                .put("index.refresh_interval", -1)
-                .build()
-        );
-        ensureYellow(INDEX_NAME);
-        final String replicaNode = internalCluster().startNode();
-        ensureGreen(INDEX_NAME);
-
-        final int initialDocCount = scaledRandomIntBetween(10, 20);
-        for (int i = 0; i < initialDocCount; i++) {
-            client().prepareIndex(INDEX_NAME).setId(String.valueOf(i)).setSource("foo", "bar").get();
-        }
-        logger.info("--> trigger round of segment replication so that replica have files");
-        refresh(INDEX_NAME);
-        waitForSearchableDocs(initialDocCount, primaryNode, replicaNode);
-        verifyStoreContent();
-        flush(INDEX_NAME);
-        waitForSegmentReplication(replicaNode);
-
-        // Corrupt one file on replica to ensure there is replica shard failure
-        final IndexShard replicaShard = getIndexShard(replicaNode, INDEX_NAME);
-        logger.info("--> Store files {}", Arrays.toString(replicaShard.store().directory().listAll()));
-        corruptRandomFile(INDEX_NAME, replicaNode);
-
-        // Insert more docs and trigger another round of segment replication
-        for (int i = initialDocCount; i < 2 * initialDocCount; i++) {
-            client().prepareIndex(INDEX_NAME).setId(String.valueOf(i)).setSource("foo", "bar").get();
-        }
-        logger.info("--> trigger another round of segment replication");
-        refresh(INDEX_NAME);
-        waitForSegmentReplication(replicaNode);
-        waitForSearchableDocs(initialDocCount, primaryNode, replicaNode);
-        verifyStoreContent();
     }
 
     public void testCancellation() throws Exception {

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
@@ -13,7 +13,9 @@ import org.opensearch.action.support.WriteRequest;
 import org.opensearch.action.update.UpdateResponse;
 import org.opensearch.client.Requests;
 import org.opensearch.cluster.health.ClusterHealthStatus;
+import org.opensearch.cluster.ClusterState;
 import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.node.DiscoveryNode;
 import org.opensearch.cluster.routing.ShardRouting;
 import org.opensearch.cluster.routing.allocation.command.CancelAllocationCommand;
 import org.opensearch.common.settings.Settings;
@@ -21,6 +23,7 @@ import org.opensearch.index.IndexModule;
 import org.opensearch.index.shard.IndexShard;
 import org.opensearch.indices.recovery.FileChunkRequest;
 import org.opensearch.indices.replication.common.ReplicationType;
+import org.opensearch.node.NodeClosedException;
 import org.opensearch.test.BackgroundIndexer;
 import org.opensearch.test.InternalTestCluster;
 import org.opensearch.test.OpenSearchIntegTestCase;
@@ -305,6 +308,57 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
 
         logger.info("--> waiting for allocation to have shards assigned");
         waitForRelocation(ClusterHealthStatus.GREEN);
+    }
+
+    /**
+     * This test validates the primary node drop does not result in shard failure on replica.
+     * @throws Exception
+     */
+    public void testNodeDropWithOngoingReplication() throws Exception {
+        internalCluster().startClusterManagerOnlyNode();
+        final String primaryNode = internalCluster().startNode();
+        createIndex(INDEX_NAME, Settings.builder().put(indexSettings()).put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 1).build());
+        ensureYellow(INDEX_NAME);
+        final String replicaNode = internalCluster().startNode();
+        ensureGreen(INDEX_NAME);
+
+        // block auto refreshes
+        assertAcked(prepareCreate("test").setSettings(Settings.builder().put("index.refresh_interval", -1)));
+        ClusterState state = client().admin().cluster().prepareState().execute().actionGet().getState();
+        DiscoveryNode primaryDiscovery = state.nodes().resolveNode(primaryNode);
+
+        CountDownLatch blockFileCopy = new CountDownLatch(1);
+        MockTransportService primaryTransportService = ((MockTransportService) internalCluster().getInstance(
+            TransportService.class,
+            primaryNode
+        ));
+        primaryTransportService.addSendBehavior(
+            internalCluster().getInstance(TransportService.class, replicaNode),
+            (connection, requestId, action, request, options) -> {
+                if (action.equals(SegmentReplicationTargetService.Actions.FILE_CHUNK)) {
+                    FileChunkRequest req = (FileChunkRequest) request;
+                    logger.debug("file chunk [{}] lastChunk: {}", req, req.lastChunk());
+                    if (req.name().endsWith("cfs") && req.lastChunk()) {
+                        try {
+                            blockFileCopy.await();
+                        } catch (InterruptedException e) {
+                            throw new RuntimeException(e);
+                        }
+                        throw new NodeClosedException(primaryDiscovery);
+                    }
+                }
+                connection.sendRequest(requestId, action, request, options);
+            }
+        );
+        final int docCount = scaledRandomIntBetween(10, 200);
+        for (int i = 0; i < docCount; i++) {
+            client().prepareIndex(INDEX_NAME).setId(Integer.toString(i)).setSource("field", "value" + i).execute().get();
+        }
+        // Refresh, this should trigger round of segment replication
+        refresh(INDEX_NAME);
+        blockFileCopy.countDown();
+        internalCluster().stopRandomNode(InternalTestCluster.nameFilter(primaryNode));
+        assertBusy(() -> { assertDocCounts(docCount, replicaNode); });
     }
 
     public void testCancellation() throws Exception {

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
@@ -49,6 +49,7 @@ import java.util.List;
 import java.util.Set;
 import java.util.TreeSet;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
 
 import static java.util.Arrays.asList;
 import static org.opensearch.action.admin.cluster.node.stats.NodesStatsRequest.Metric.FS;
@@ -377,56 +378,7 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
         assertBusy(() -> { assertDocCounts(docCount, replicaNode); });
     }
 
-    protected ShardRouting corruptRandomFile(String indexName, String nodeId) throws IOException {
-//        ClusterState state = client().admin().cluster().prepareState().get().getState();
-//        Index test = state.metadata().index("test").getIndex();
-//        GroupShardsIterator shardIterators = state.getRoutingTable().activePrimaryShardsGrouped(new String[] { "test" }, false);
-//        List<ShardIterator> iterators = iterableAsArrayList(shardIterators);
-//        ShardIterator shardIterator = RandomPicks.randomFrom(random(), iterators);
-//        ShardRouting shardRouting = shardIterator.nextOrNull();
-//        assertNotNull(shardRouting);
-//        assertTrue(shardRouting.primary());
-//        assertTrue(shardRouting.assignedToNode());
-//        String nodeId = shardRouting.currentNodeId();
-        ClusterState state = client().admin().cluster().prepareState().get().getState();
-        Index test = state.metadata().index(indexName).getIndex();
-
-        final IndexShard indexShard = getIndexShard(nodeId, INDEX_NAME);
-        final ShardRouting shardRouting = indexShard.routingEntry();
-
-        NodesStatsResponse nodeStatses = client().admin().cluster().prepareNodesStats(nodeId).addMetric(FS.metricName()).get();
-        Set<Path> files = new TreeSet<>(); // treeset makes sure iteration order is deterministic
-        for (FsInfo.Path info : nodeStatses.getNodes().get(0).getFs()) {
-            String path = info.getPath();
-            Path file = PathUtils.get(path)
-                .resolve("indices")
-                .resolve(test.getUUID())
-                .resolve(Integer.toString(shardRouting.getId()))
-                .resolve("index");
-            if (Files.exists(file)) { // multi data path might only have one path in use
-                try (Directory dir = FSDirectory.open(file)) {
-                    SegmentInfos segmentCommitInfos = Lucene.readSegmentInfos(dir);
-//                    if (includePerCommitFiles) {
-//                        files.add(file.resolve(segmentCommitInfos.getSegmentsFileName()));
-//                    }
-                    for (SegmentCommitInfo commitInfo : segmentCommitInfos) {
-                        if (commitInfo.getDelCount() + commitInfo.getSoftDelCount() == commitInfo.info.maxDoc()) {
-                            // don't corrupt fully deleted segments - they might be removed on snapshot
-                            continue;
-                        }
-                        for (String commitFile : commitInfo.files()) {
-//                            if (includePerCommitFiles || isPerSegmentFile(commitFile)) {
-//                                files.add(file.resolve(commitFile));
-//                            }
-                            logger.info("--> File name {}", commitFile);
-                        }
-                    }
-                }
-            }
-        }
-        CorruptionUtils.corruptFile(random(), files.toArray(new Path[0]));
-        return shardRouting;
-    }
+    @AwaitsFix(bugUrl = "https://github.com/opensearch-project/OpenSearch/issues/6578")
     public void testFileCorruption() throws Exception {
         internalCluster().startClusterManagerOnlyNode();
         final String primaryNode = internalCluster().startNode();
@@ -436,30 +388,31 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
         final String replicaNode = internalCluster().startNode();
         ensureGreen(INDEX_NAME);
 
-        final int initialDocCount = scaledRandomIntBetween(10, 200);
+        final int initialDocCount = scaledRandomIntBetween(10, 20);
         for (int i = 0; i < initialDocCount; i++) {
             client().prepareIndex(INDEX_NAME).setId(String.valueOf(i)).setSource("foo", "bar").get();
         }
         logger.info("--> trigger round of segment replication so that replica have files");
-        flush(INDEX_NAME);
+        refresh(INDEX_NAME);
         waitForSearchableDocs(initialDocCount, primaryNode, replicaNode);
         verifyStoreContent();
+        flush(INDEX_NAME);
+        waitForSegmentReplication(replicaNode);
 
-        // Primary & replica should be in sync now. Corrupt one file on replica to ensure there is replica shard failure
-        final IndexShard primaryShard = getIndexShard(primaryNode, INDEX_NAME);
-        logger.info("--> Store files {}", Arrays.toString(primaryShard.store().directory().listAll()));
-//        final ShardRouting replicaShardRouting = replicaShard.routingEntry();
+        // Corrupt one file on replica to ensure there is replica shard failure
+        final IndexShard replicaShard = getIndexShard(replicaNode, INDEX_NAME);
+        logger.info("--> Store files {}", Arrays.toString(replicaShard.store().directory().listAll()));
+        corruptRandomFile(INDEX_NAME, replicaNode);
 
-//        corruptRandomFile(INDEX_NAME, replicaNode);
-//
-//        // Insert more docs and trigger another round of segment replication
-//        for (int i = initialDocCount; i < 2 * initialDocCount; i++) {
-//            client().prepareIndex(INDEX_NAME).setId(String.valueOf(i)).setSource("foo", "bar").get();
-//        }
-//        logger.info("--> refresh again and expect shard failure on replica");
-//        refresh(INDEX_NAME);
-
-        // Should see segrep failure
+        // Insert more docs and trigger another round of segment replication
+        for (int i = initialDocCount; i < 2 * initialDocCount; i++) {
+            client().prepareIndex(INDEX_NAME).setId(String.valueOf(i)).setSource("foo", "bar").get();
+        }
+        logger.info("--> trigger another round of segment replication");
+        refresh(INDEX_NAME);
+        waitForSegmentReplication(replicaNode);
+        waitForSearchableDocs(initialDocCount, primaryNode, replicaNode);
+        verifyStoreContent();
     }
 
     public void testCancellation() throws Exception {

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
@@ -17,6 +17,7 @@ import org.opensearch.cluster.ClusterState;
 import org.opensearch.cluster.metadata.IndexMetadata;
 import org.opensearch.cluster.node.DiscoveryNode;
 import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.cluster.routing.ShardRoutingState;
 import org.opensearch.cluster.routing.allocation.command.CancelAllocationCommand;
 import org.opensearch.common.settings.Settings;
 import org.opensearch.index.IndexModule;
@@ -317,14 +318,31 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
     public void testNodeDropWithOngoingReplication() throws Exception {
         internalCluster().startClusterManagerOnlyNode();
         final String primaryNode = internalCluster().startNode();
-        createIndex(INDEX_NAME, Settings.builder().put(indexSettings()).put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 1).build());
+        createIndex(
+            INDEX_NAME,
+            Settings.builder()
+                .put(indexSettings())
+                .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 1)
+                .build()
+        );
         ensureYellow(INDEX_NAME);
         final String replicaNode = internalCluster().startNode();
         ensureGreen(INDEX_NAME);
 
         // block auto refreshes
-        assertAcked(prepareCreate("test").setSettings(Settings.builder().put("index.refresh_interval", -1)));
+        assertAcked(prepareCreate(INDEX_NAME).setSettings(Settings.builder().put("index.refresh_interval", -1)));
         ClusterState state = client().admin().cluster().prepareState().execute().actionGet().getState();
+        // Get replica allocation id
+        final String replicaAllocationId = state.routingTable()
+            .index(INDEX_NAME)
+            .shardsWithState(ShardRoutingState.STARTED)
+            .stream()
+            .filter(routing -> routing.primary() == false)
+            .findFirst()
+            .get()
+            .allocationId()
+            .getId();
         DiscoveryNode primaryDiscovery = state.nodes().resolveNode(primaryNode);
 
         CountDownLatch blockFileCopy = new CountDownLatch(1);
@@ -359,6 +377,18 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
         blockFileCopy.countDown();
         internalCluster().stopRandomNode(InternalTestCluster.nameFilter(primaryNode));
         assertBusy(() -> { assertDocCounts(docCount, replicaNode); });
+        state = client().admin().cluster().prepareState().execute().actionGet().getState();
+        // replica now promoted as primary should have same allocation id
+        final String currentAllocationID = state.routingTable()
+            .index(INDEX_NAME)
+            .shardsWithState(ShardRoutingState.STARTED)
+            .stream()
+            .filter(routing -> routing.primary())
+            .findFirst()
+            .get()
+            .allocationId()
+            .getId();
+        assertEquals(currentAllocationID, replicaAllocationId);
     }
 
     public void testCancellation() throws Exception {

--- a/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java
+++ b/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java
@@ -172,13 +172,11 @@ public class SegmentReplicationTarget extends ReplicationTarget {
         finalizeListener.whenComplete(r -> listener.onResponse(null), listener::onFailure);
     }
 
-    private void getFiles(CheckpointInfoResponse checkpointInfo, StepListener<GetSegmentFilesResponse> getFilesListener) throws IOException {
+    private void getFiles(CheckpointInfoResponse checkpointInfo, StepListener<GetSegmentFilesResponse> getFilesListener)
+        throws IOException {
         cancellableThreads.checkForCancel();
         state.setStage(SegmentReplicationState.Stage.FILE_DIFF);
-        final Store.RecoveryDiff diff = Store.segmentReplicationDiff(
-            checkpointInfo.getMetadataMap(),
-            indexShard.getSegmentMetadataMap()
-        );
+        final Store.RecoveryDiff diff = Store.segmentReplicationDiff(checkpointInfo.getMetadataMap(), indexShard.getSegmentMetadataMap());
         logger.trace("Replication diff for checkpoint {} {}", checkpointInfo.getCheckpoint(), diff);
         /*
          * Segments are immutable. So if the replica has any segments with the same name that differ from the one in the incoming

--- a/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java
+++ b/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java
@@ -172,46 +172,43 @@ public class SegmentReplicationTarget extends ReplicationTarget {
         finalizeListener.whenComplete(r -> listener.onResponse(null), listener::onFailure);
     }
 
-    private void getFiles(CheckpointInfoResponse checkpointInfo, StepListener<GetSegmentFilesResponse> getFilesListener) {
-        ActionListener.completeWith(getFilesListener, () -> {
-            cancellableThreads.checkForCancel();
-            state.setStage(SegmentReplicationState.Stage.FILE_DIFF);
-            final Store.RecoveryDiff diff = Store.segmentReplicationDiff(
-                checkpointInfo.getMetadataMap(),
-                indexShard.getSegmentMetadataMap()
-            );
-            logger.trace("Replication diff for checkpoint {} {}", checkpointInfo.getCheckpoint(), diff);
-            /*
-             * Segments are immutable. So if the replica has any segments with the same name that differ from the one in the incoming
-             * snapshot from source that means the local copy of the segment has been corrupted/changed in some way and we throw an
-             * IllegalStateException to fail the shard
-             */
-            if (diff.different.isEmpty() == false) {
-                IllegalStateException illegalStateException = new IllegalStateException(
-                    new ParameterizedMessage(
-                        "Shard {} has local copies of segments that differ from the primary {}",
-                        indexShard.shardId(),
-                        diff.different
-                    ).getFormattedMessage()
-                );
-                ReplicationFailedException rfe = new ReplicationFailedException(
+    private void getFiles(CheckpointInfoResponse checkpointInfo, StepListener<GetSegmentFilesResponse> getFilesListener) throws IOException {
+        cancellableThreads.checkForCancel();
+        state.setStage(SegmentReplicationState.Stage.FILE_DIFF);
+        final Store.RecoveryDiff diff = Store.segmentReplicationDiff(
+            checkpointInfo.getMetadataMap(),
+            indexShard.getSegmentMetadataMap()
+        );
+        logger.trace("Replication diff for checkpoint {} {}", checkpointInfo.getCheckpoint(), diff);
+        /*
+         * Segments are immutable. So if the replica has any segments with the same name that differ from the one in the incoming
+         * snapshot from source that means the local copy of the segment has been corrupted/changed in some way and we throw an
+         * IllegalStateException to fail the shard
+         */
+        if (diff.different.isEmpty() == false) {
+            IllegalStateException illegalStateException = new IllegalStateException(
+                new ParameterizedMessage(
+                    "Shard {} has local copies of segments that differ from the primary {}",
                     indexShard.shardId(),
-                    "different segment files",
-                    illegalStateException
-                );
-                fail(rfe, true);
-                throw rfe;
-            }
+                    diff.different
+                ).getFormattedMessage()
+            );
+            ReplicationFailedException rfe = new ReplicationFailedException(
+                indexShard.shardId(),
+                "different segment files",
+                illegalStateException
+            );
+            fail(rfe, true);
+            throw rfe;
+        }
 
-            for (StoreFileMetadata file : diff.missing) {
-                state.getIndex().addFileDetail(file.name(), file.length(), false);
-            }
-            // always send a req even if not fetching files so the primary can clear the copyState for this shard.
-            state.setStage(SegmentReplicationState.Stage.GET_FILES);
-            cancellableThreads.checkForCancel();
-            source.getSegmentFiles(getId(), checkpointInfo.getCheckpoint(), diff.missing, store, getFilesListener);
-            return null;
-        });
+        for (StoreFileMetadata file : diff.missing) {
+            state.getIndex().addFileDetail(file.name(), file.length(), false);
+        }
+        // always send a req even if not fetching files so the primary can clear the copyState for this shard.
+        state.setStage(SegmentReplicationState.Stage.GET_FILES);
+        cancellableThreads.checkForCancel();
+        source.getSegmentFiles(getId(), checkpointInfo.getCheckpoint(), diff.missing, store, getFilesListener);
     }
 
     private void finalizeReplication(CheckpointInfoResponse checkpointInfoResponse, ActionListener<Void> listener) {

--- a/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java
+++ b/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java
@@ -108,7 +108,7 @@ public class SegmentReplicationTarget extends ReplicationTarget {
 
     @Override
     public void notifyListener(ReplicationFailedException e, boolean sendShardFailure) {
-        // Cancellations still are passed to our SegmentReplicationListner as failures, if we have failed because of cancellation
+        // Cancellations still are passed to our SegmentReplicationListener as failures, if we have failed because of cancellation
         // update the stage.
         final Throwable cancelledException = ExceptionsHelper.unwrap(e, CancellableThreads.ExecutionCancelledException.class);
         if (cancelledException != null) {
@@ -172,50 +172,56 @@ public class SegmentReplicationTarget extends ReplicationTarget {
         finalizeListener.whenComplete(r -> listener.onResponse(null), listener::onFailure);
     }
 
-    private void getFiles(CheckpointInfoResponse checkpointInfo, StepListener<GetSegmentFilesResponse> getFilesListener)
-        throws IOException {
-        cancellableThreads.checkForCancel();
-        state.setStage(SegmentReplicationState.Stage.FILE_DIFF);
-        final Store.RecoveryDiff diff = Store.segmentReplicationDiff(checkpointInfo.getMetadataMap(), indexShard.getSegmentMetadataMap());
-        logger.trace("Replication diff for checkpoint {} {}", checkpointInfo.getCheckpoint(), diff);
-        /*
-         * Segments are immutable. So if the replica has any segments with the same name that differ from the one in the incoming
-         * snapshot from source that means the local copy of the segment has been corrupted/changed in some way and we throw an
-         * IllegalStateException to fail the shard
-         */
-        if (diff.different.isEmpty() == false) {
-            IllegalStateException illegalStateException = new IllegalStateException(
-                new ParameterizedMessage(
-                    "Shard {} has local copies of segments that differ from the primary {}",
+    private void getFiles(CheckpointInfoResponse checkpointInfo, StepListener<GetSegmentFilesResponse> getFilesListener) {
+        ActionListener.completeWith(getFilesListener, () -> {
+            cancellableThreads.checkForCancel();
+            state.setStage(SegmentReplicationState.Stage.FILE_DIFF);
+            final Store.RecoveryDiff diff = Store.segmentReplicationDiff(
+                checkpointInfo.getMetadataMap(),
+                indexShard.getSegmentMetadataMap()
+            );
+            logger.trace("Replication diff for checkpoint {} {}", checkpointInfo.getCheckpoint(), diff);
+            /*
+             * Segments are immutable. So if the replica has any segments with the same name that differ from the one in the incoming
+             * snapshot from source that means the local copy of the segment has been corrupted/changed in some way and we throw an
+             * IllegalStateException to fail the shard
+             */
+            if (diff.different.isEmpty() == false) {
+                IllegalStateException illegalStateException = new IllegalStateException(
+                    new ParameterizedMessage(
+                        "Shard {} has local copies of segments that differ from the primary {}",
+                        indexShard.shardId(),
+                        diff.different
+                    ).getFormattedMessage()
+                );
+                ReplicationFailedException rfe = new ReplicationFailedException(
                     indexShard.shardId(),
-                    diff.different
-                ).getFormattedMessage()
-            );
-            ReplicationFailedException rfe = new ReplicationFailedException(
-                indexShard.shardId(),
-                "replication failure due to different segment files",
-                illegalStateException
-            );
-            fail(rfe, true);
-        }
+                    "different segment files",
+                    illegalStateException
+                );
+                fail(rfe, true);
+                throw rfe;
+            }
 
-        for (StoreFileMetadata file : diff.missing) {
-            state.getIndex().addFileDetail(file.name(), file.length(), false);
-        }
-        // always send a req even if not fetching files so the primary can clear the copyState for this shard.
-        state.setStage(SegmentReplicationState.Stage.GET_FILES);
-        cancellableThreads.checkForCancel();
-        source.getSegmentFiles(getId(), checkpointInfo.getCheckpoint(), diff.missing, store, getFilesListener);
+            for (StoreFileMetadata file : diff.missing) {
+                state.getIndex().addFileDetail(file.name(), file.length(), false);
+            }
+            // always send a req even if not fetching files so the primary can clear the copyState for this shard.
+            state.setStage(SegmentReplicationState.Stage.GET_FILES);
+            cancellableThreads.checkForCancel();
+            source.getSegmentFiles(getId(), checkpointInfo.getCheckpoint(), diff.missing, store, getFilesListener);
+            return null;
+        });
     }
 
     private void finalizeReplication(CheckpointInfoResponse checkpointInfoResponse, ActionListener<Void> listener) {
         ActionListener.completeWith(listener, () -> {
             cancellableThreads.checkForCancel();
             state.setStage(SegmentReplicationState.Stage.FINALIZE_REPLICATION);
-            multiFileWriter.renameAllTempFiles();
-            final Store store = store();
-            store.incRef();
             try {
+                multiFileWriter.renameAllTempFiles();
+                final Store store = store();
+                store.incRef();
                 // Deserialize the new SegmentInfos object sent from the primary.
                 final ReplicationCheckpoint responseCheckpoint = checkpointInfoResponse.getCheckpoint();
                 SegmentInfos infos = SegmentInfos.readCommit(

--- a/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java
+++ b/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java
@@ -184,15 +184,19 @@ public class SegmentReplicationTarget extends ReplicationTarget {
          * IllegalStateException to fail the shard
          */
         if (diff.different.isEmpty() == false) {
-            getFilesListener.onFailure(
-                new IllegalStateException(
-                    new ParameterizedMessage(
-                        "Shard {} has local copies of segments that differ from the primary {}",
-                        indexShard.shardId(),
-                        diff.different
-                    ).getFormattedMessage()
-                )
+            IllegalStateException illegalStateException = new IllegalStateException(
+                new ParameterizedMessage(
+                    "Shard {} has local copies of segments that differ from the primary {}",
+                    indexShard.shardId(),
+                    diff.different
+                ).getFormattedMessage()
             );
+            ReplicationFailedException rfe = new ReplicationFailedException(
+                indexShard.shardId(),
+                "replication failure due to different segment files",
+                illegalStateException
+            );
+            fail(rfe, true);
         }
 
         for (StoreFileMetadata file : diff.missing) {

--- a/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java
+++ b/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java
@@ -362,7 +362,7 @@ public class SegmentReplicationTargetService implements IndexEventListener {
                             completedReplications.put(target.shardId(), target);
                         }
                     } else {
-                        onGoingReplications.fail(replicationId, new ReplicationFailedException("Segment Replication failed", e), true);
+                        onGoingReplications.fail(replicationId, new ReplicationFailedException("Segment Replication failed", e), false);
                     }
                 }
             });

--- a/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java
+++ b/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java
@@ -352,7 +352,6 @@ public class SegmentReplicationTargetService implements IndexEventListener {
                 @Override
                 public void onFailure(Exception e) {
                     Throwable cause = ExceptionsHelper.unwrapCause(e);
-                    logger.warn("Replication failure", e);
                     if (cause instanceof CancellableThreads.ExecutionCancelledException) {
                         if (onGoingReplications.getTarget(replicationId) != null) {
                             IndexShard indexShard = onGoingReplications.getTarget(replicationId).indexShard();
@@ -362,6 +361,8 @@ public class SegmentReplicationTargetService implements IndexEventListener {
                             onGoingReplications.fail(replicationId, new ReplicationFailedException(indexShard, cause), false);
                             completedReplications.put(target.shardId(), target);
                         }
+                    } else {
+                        onGoingReplications.fail(replicationId, new ReplicationFailedException("Segment Replication failed", e), true);
                     }
                 }
             });

--- a/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java
+++ b/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java
@@ -352,6 +352,7 @@ public class SegmentReplicationTargetService implements IndexEventListener {
                 @Override
                 public void onFailure(Exception e) {
                     Throwable cause = ExceptionsHelper.unwrapCause(e);
+                    logger.warn("Replication failure", e);
                     if (cause instanceof CancellableThreads.ExecutionCancelledException) {
                         if (onGoingReplications.getTarget(replicationId) != null) {
                             IndexShard indexShard = onGoingReplications.getTarget(replicationId).indexShard();
@@ -361,8 +362,6 @@ public class SegmentReplicationTargetService implements IndexEventListener {
                             onGoingReplications.fail(replicationId, new ReplicationFailedException(indexShard, cause), false);
                             completedReplications.put(target.shardId(), target);
                         }
-                    } else {
-                        onGoingReplications.fail(replicationId, new ReplicationFailedException("Segment Replication failed", e), true);
                     }
                 }
             });

--- a/server/src/test/java/org/opensearch/indices/replication/SegmentReplicationTargetTests.java
+++ b/server/src/test/java/org/opensearch/indices/replication/SegmentReplicationTargetTests.java
@@ -372,8 +372,8 @@ public class SegmentReplicationTargetTests extends IndexShardTestCase {
 
             @Override
             public void onFailure(Exception e) {
-                assert (e instanceof IllegalStateException);
-                segrepTarget.fail(new ReplicationFailedException(e), false);
+                assert (e instanceof ReplicationFailedException);
+                assert (e.getMessage().contains("different segment files"));
             }
         });
     }

